model-server/src/test/kotlin/org/modelix/model/server/ReplicatedModelTest.kt

```diff
@@ -0,0 +1,165 @@
+import org.modelix.model.client2.ReplicatedModel
+import org.modelix.model.lazy.BranchReference
+import org.modelix.model.lazy.CLTree
+import org.modelix.model.lazy.CLVersion
+import org.modelix.model.lazy.RepositoryId
+import org.modelix.model.operations.OTBranch
+import org.modelix.model.server.handlers.ModelReplicationServer
+import org.modelix.model.server.store.InMemoryStoreClient
+import org.modelix.model.server.store.forContextRepository
+import java.util.UUID
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+class ReplicatedModelTest {
+
+    @Test
+    fun startsWithLatestVersion() = runTest {
+        val client = createModelClient()
+        val repositoryId = RepositoryId(UUID.randomUUID().toString())
+
+        // Step 1: prepare repository with two versions beside the initial version
+        // Step 1.1: create an empty repository
+        val initialVersion = client.initRepository(repositoryId) as CLVersion
+
+        // Step 1.2: add a new child node to get a new version
+        val defaultBranchReference = repositoryId.getBranchReference()
+        addHelloChild(initialVersion, client, defaultBranchReference)
+
+        // Step 2: in a new client, fetch the latest repository data
+        val newClient = createModelClient()
+        // we do not provide an initial version, so we expect to fetch the latest one (with one "hello" child)
+        val replicatedModel = ReplicatedModel(newClient, defaultBranchReference)
+        val branch = replicatedModel.start { branch ->
+            // Step 3: check, here we must have 1 "hello" child
+            val children = getHelloChildrenOfRootNode(branch)
+            assertEquals(1, children.size)
+        }
+
+        // Step 5: check, eventually we must have the one "hello" child
+        val children = getHelloChildrenOfRootNode(branch)
+        assertEquals(1, children.size)
+    }
+
+    @Test
+    fun startsWithSpecificVersion() = runTest {
+        val client = createModelClient()
+        val repositoryId = RepositoryId(UUID.randomUUID().toString())
+        val defaultBranchReference = repositoryId.getBranchReference()
+
+        // Step 1: prepare repository with two versions beside the initial version
+        // Step 1.1: create an empty repository
+        val initialVersion = client.initRepository(repositoryId) as CLVersion
+
+        // Step 1.2: add a new child node to get a new version
+        addHelloChild(initialVersion, client, defaultBranchReference)
+
+        // Step 2: in a new client, fetch the oneHelloChildVersion
+        val newClient = createModelClient()
+
+        // Step 2.1: to avoid version was not created by this client exception
+        val initialVersionClone = newClient.loadVersion(repositoryId, initialVersion.getContentHash(), null)
+
+        val scope = CoroutineScope(Dispatchers.Default)
```

Comment:
Why is this used now?

Reply:
Because there is no direct way to check if ReplicatedModel has fetched the latest version. Indirectly, I have to wait until the "hello" child appears. 

In order to avoid a fixed wait of 500 ms in the coroutine that is running at line 97, I create a CountdownLatch and start a new coroutine that checks if the "hello" child has arrived. If it has, then it opens the latch. The coroutine running at line 97 will be stopped by the Latch and wait until the Latch is opened or 500 ms has passed whichever occurs first.

Reply:
ah, and the single-threaded test dispatcher would be blocked already by waiting on the latch?